numpy/lib/function_base.py

```diff
@@ -1323,21 +1323,15 @@ def piecewise(x, condlist, funclist, *args, **kw):
     """
     x = asanyarray(x)
     n2 = len(funclist)
-    if (isscalar(condlist) or not (isinstance(condlist[0], list) or
-                                   isinstance(condlist[0], ndarray))):
-        if not isscalar(condlist) and x.size == 1 and x.ndim == 0:
-            condlist = [[c] for c in condlist]
-        else:
-            condlist = [condlist]
+
+    # undocumented: single condition is promoted to a list of one condition
+    if np.ndim(condlist) == 0 or (
```

Comment:
This took me very long to parse; could you add a comment for what the second part of the if statement is for? It might have helped me a little to have seen a reversed order `x.ndim >0 and not isinstance(...)`

Reply:
So, part of the problem is that this statement is to hit an arbitrary selection of cases where the user maybe forgot to pass a list. It performs the following mappings

* `piecewise(x_1d, cond_1d, ...)` -> `piecewise(x_1d, [cond_1d])`
* `piecewise(x_0d, cond_0d, ...)` -> `piecewise(x_1d, [cond_0d])`  (after the second commit)

But doesn't do anything to

*  `piecewise(x_2d, cond_2d, ...)`-
*  `piecewise(x_3d, cond_3d, ...)`

Which it interprets as broadcasting an `n-1`d condition against an `nd` array.

This is super inconsistent - perhaps I should leave the `isscalar`, and just deprecate the entire code path?